lnrpc/routerrpc/router_server.go

```diff
@@ -398,14 +445,399 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += int64(hopFee)
+
+	// Use the hop hint's cltv delta as the payment request's final cltv
+	// delta. The actual final cltv delta of the invoice will be added to
+	// the payment probe's cltv delta.
+	probeRequest.FinalCltvDelta = int32(lspHop.CLTVExpiryDelta)
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to fee the
+	// payment probe returned.
+	resp.RoutingFeeMsat += int64(hopFee)
+
+	// Add the final cltv delta of the invoice to the payment probe's total
+	// cltv delta. This is the cltv delta for the hop behind the LSP.
+	resp.TimeLockDelay += int64(payReq.MinFinalCLTVExpiry())
+
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to heuristically determine a worst-case
+// estimate.
+func isLSP(routeHints [][]zpay32.HopHint) bool {
+	if len(routeHints) == 0 || len(routeHints[0]) == 0 {
+		return false
+	}
+
+	refNodeID := routeHints[0][len(routeHints[0])-1].NodeID
+	for i := 1; i < len(routeHints); i++ {
+		// Skip empty route hints.
+		if len(routeHints[i]) == 0 {
+			continue
+		}
+
+		lastHop := routeHints[i][len(routeHints[i])-1]
+		idMatchesRefNode := bytes.Equal(
+			lastHop.NodeID.SerializeCompressed(),
+			refNodeID.SerializeCompressed(),
+		)
+		if !idMatchesRefNode {
+			return false
+		}
+	}
+
+	return true
+}
+
+// adjustedRouteHints contains the route hints that allow a caller to probe an
+// LSP.
+type adjustedRouteHints struct {
+	// adjustedRouteHints are the route hints that allow the caller to probe
+	// the LSP.
+	hints [][]zpay32.HopHint
+
+	// lspHop is the hop hint that identifies the LSP node.
+	lspHop *zpay32.HopHint
+}
+
+// prepareLspRouteHints assumes that the isLsp heuristic returned true for the
+// route hints passed in here. It constructs a modified list of route hints that
+// allows the caller to probe the LSP, which itself is returned as a separate
+// hop hint.
+func prepareLspRouteHints(routeHints [][]zpay32.HopHint,
+	amt lnwire.MilliSatoshi) *adjustedRouteHints {
+
+	// Create the LSP hop hint. We are probing for the worst case fee and
+	// cltv delta. So we look for the max values amongst all LSP hop hints.
+	refHint := routeHints[0][len(routeHints[0])-1]
```

Comment:
I don't think we need to return an error but `false`? If the hints are `nil` they will be safely processed by `sendPayment`. That's tested in an itest `probe based estimate, empty route hint`.

Reply:
right, `isLSP` would handle this, but the question is where do you draw the line of checking for nil slices if you access those (or other situations where panics could happen). for me it would be within the scope of a function, as it's a reusable piece of code.

Reply:
apologies, I thought the mentioned line is in `isLsp` but it is in `prepareLspRouteHints`. Agree with what your are saying, will change.